x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeProcess.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.process;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.ZonedDateTime;
+
+/**
+ * Interface representing a native C++ process
+ */
+public interface NativeProcess extends Closeable {
+
+    /**
+     * Is the process ready to receive data?
+     * @return {@code true} if the process is ready to receive data
+     */
+    boolean isReady();
+
+    /**
+     * Write the record to the process. The record parameter should not be encoded
+     * (i.e. length encoded) the implementation will apply the correct encoding.
+     *
+     * @param record Plain array of strings, implementors of this class should
+     *               encode the record appropriately
+     * @throws IOException If the write failed
+     */
+    void writeRecord(String[] record) throws IOException;
+
+    /**
+     * Ask the process to persist its state in the background
+     * @throws IOException If writing the request fails
+     */
+    void persistJob() throws IOException;
```

Comment:
nit: trying to avoid our profligate and overloaded use of `job` perhaps rename to `persistState`